### DISCUSSION TITLE
Does Cypress/puppeteer plugin support userFlow in lighthouse?

### DISCUSSION THREAD

**Discussion by yasserhussien** [cypress-io/cypress#30041] [Questions and Help]
right now I am trying to use lighthouse with cypress and I read that the user flow in lighthouse is only integrated with puppeteer so I tried to use cypress/puppeteer plugin but it doesn't work for me because the function startFlow from lighthouse assume that he will receive the page from lighthouse/types/puppeteer but it received from puppeteer-core/lib/types.
is there a way to solve this problem ?

**Comment by portyu9** [comment]
Hi @yasserhussien 

Yes, but I would solve the type mismatch by making Lighthouse and `@cypress/puppeteer` use the **same Puppeteer installation/version**, rather than casting the `Page`.

The current `@cypress/puppeteer` API supports supplying your own `puppeteer-core` implementation:

```ts
import { defineConfig } from 'cypress'
import { setup } from '@cypress/puppeteer'
import puppeteer from 'puppeteer-core'

export default defineConfig({
  e2e: {
    setupNodeEvents(on) {
      setup({
        on,
        puppeteer,
        onMessage: {
          async myHandler(browser) {
            // `browser` and pages created/retrieved from it now come
            // from the Puppeteer implementation supplied above.
          },
        },
      })
    },
  },
})
```

Then make sure the `puppeteer-core` version you install is compatible with the Lighthouse version in the project:

```sh
npm ls lighthouse puppeteer puppeteer-core
```

The error you described usually means TypeScript is seeing `Page` types from two different Puppeteer module/version trees. Those types are not necessarily assignable even though both are called `Page`.

I would **not** fix that with:

```ts
page as SomeOtherPuppeteerPage
```

because that only hides the type error and does not guarantee runtime compatibility.

There is one additional limitation to keep in mind: Lighthouse `startFlow()` requires a real Puppeteer `Page`. `@cypress/puppeteer` exposes the Cypress-launched browser, but Cypress's normal AUT is rendered inside the Cypress runner. If the `Page` you obtain is the Cypress runner rather than the document you intend to audit, aligning the types alone will not make the Lighthouse flow target the AUT correctly.

So the supported pieces are available today, but I would first:

1. use one compatible `puppeteer-core` version for both integrations;
2. pass that implementation through the plugin's `puppeteer` option;
3. verify that the `Page` supplied to `startFlow()` is actually the page you intend Lighthouse to audit.

If you specifically need Lighthouse navigation/timespan/snapshot flows over an application page rather than a second tab/window, a standalone Lighthouse + Puppeteer flow may still be simpler than coupling that flow to Cypress.

No screenshots or attachments were included in this discussion.

Please mark this answer as accepted if it helped, thank you.
